src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -2178,9 +2179,93 @@ static public TimeZoneInfo FindSystemTimeZoneById(string id)
             return result;
         }
 
-        static public ReadOnlyCollection<TimeZoneInfo> GetSystemTimeZones() {
-            // UNIXTODO
-            throw new NotImplementedException();
+        /// <summary>
+        /// Returns a collection of TimeZoneInfo instances containing all valid TimeZones
+        /// from the local machine.  The entries in the collection are sorted by
+        /// 'DisplayName'.
+        ///
+        /// This method does *not* throw TimeZoneNotFoundException or
+        /// InvalidTimeZoneException, but it can throw a SecurityException.
+        /// </summary>
+        static public ReadOnlyCollection<TimeZoneInfo> GetSystemTimeZones()
+        {
+            // UNIXTODO: Issue #2497 use CachedData
+
+            List<TimeZoneInfo> timeZoneInfos = new List<TimeZoneInfo>();
+
+            string timeZoneDirectory = GetTimeZoneDirectory();
+            foreach (string timeZoneId in GetTimeZoneIds(timeZoneDirectory))
+            {
+                if (!string.IsNullOrEmpty(timeZoneId))
+                {
+                    try
+                    {
+                        TimeZoneInfo timeZoneInfo = FindSystemTimeZoneById(timeZoneId);
+                        timeZoneInfos.Add(timeZoneInfo);
+                    }
+                    catch (InvalidTimeZoneException) { }
+                    catch (TimeZoneNotFoundException) { }
+                }
+            }
+
+            // sort and copy the TimeZoneInfo's into a ReadOnlyCollection for the user
+            timeZoneInfos.Sort(new TimeZoneInfoComparer());
+
+            return new ReadOnlyCollection<TimeZoneInfo>(timeZoneInfos);
+        }
+
+        /// <summary>
+        /// Returns a collection of TimeZone Id values from the zone.tab file in the timeZoneDirectory.
+        /// </summary>
+        /// <remarks>
+        /// Lines that start with # are comments and are skipped.
+        /// </remarks>
+        private static IEnumerable<string> GetTimeZoneIds(string timeZoneDirectory)
+        {
+            string[] zoneTabFileLines = null;
+            try
+            {
+                zoneTabFileLines = File.ReadAllLines(Path.Combine(timeZoneDirectory, c_zoneTabFileName));
+            }
+            catch (IOException) { }
```

Comment:
The [MSDN doc](https://msdn.microsoft.com/en-us/library/system.timezoneinfo.getsystemtimezones(v=vs.110).aspx) says this method can throw OutOfMemoryException and SecurityException.  And [comments in the code](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/TimeZoneInfo.cs#L952) say this method does not throw TimeZoneNotFound or InvalidTimeZone exceptions.  The Windows version of this code doesn't throw if it can't find the registry information.

Given these things, it feels incorrect to throw in the Linux case.